### PR TITLE
Fix setSubsetString return value on error

### LINKED ISSUE #34259: filter expression error returns true
Hi,

teaching QGIS on 3.4.15 today, we face a bug on MAC and windows (have to try on ubuntu) :
When we create a filter on a layer, if the expression contains an error ( for example a parenthesis not closed), the expression returns true so the layer is not filtered.
filter_error_returns_true

### PR DESCRIPTION
**Pull request by uclaros**
## Description
`QgsOgrProvider::setSubsetString` should return false on malformed sql.

Fixes #34259

### CODE REVIEW

**Review by elpaso — APPROVED** [review]

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -4368,7 +4368,10 @@ OGRLayerH QgsOgrProviderUtils::setSubsetString( OGRLayerH layer, GDALDatasetH ds
   }
   else
   {
-    OGR_L_SetAttributeFilter( layer, encoding->fromUnicode( subsetString ).constData() );
+    if ( OGR_L_SetAttributeFilter( layer, encoding->fromUnicode( subsetString ).constData() ) != OGRERR_NONE )
+    {
+      return nullptr;
```
Comment (m-kuhn) [inline src/core/providers/ogr/qgsogrprovider.cpp]: Any chance to also log an error message if this happens?

Comment (uclaros) [reply src/core/providers/ogr/qgsogrprovider.cpp]: `QgsOgrProviderUtils::setSubsetString` is just a utility method used by `QgsOgrProvider::_setSubsetString` that handles the error and the logging:
https://github.com/qgis/QGIS/blob/0b4cc5adaeed099f83a47a527acefa6e2ac8bd86/src/core/providers/ogr/qgsogrprovider.cpp#L2067-L2073
then
https://github.com/qgis/QGIS/blob/0b4cc5adaeed099f83a47a527acefa6e2ac8bd86/src/core/qgsvectordataprovider.cpp#L705-L710

So I don't think extra logging is needed. I could add some unit test though, I just need to figure out where!...

Comment (m-kuhn) [reply src/core/providers/ogr/qgsogrprovider.cpp]: That sounds good.

Some tests exist already in `tests/src/python/test_provider_ogr*.py`, you could build on top of that

Comment (uclaros) [reply src/core/providers/ogr/qgsogrprovider.cpp]: Yes, but shapefile and spatialite tests are in `test_provider_shapefile.py` and `test_provider_spatialite.py`, are they not using ogrdataprovider? I think I'll put them on `test_provider_ogr.py` that is more generic, though currently all tests regarding `setSubsetString` are in each format's test file.

Comment (uclaros) [reply src/core/providers/ogr/qgsogrprovider.cpp]: Changed my mind and added the test on `test_provider_shapefile.py` since most ogr tests are there.

**Comment by elpaso** [comment]
@uclaros any chance to see it backported to 3.10 and 3.12?  I don't see any potential regression unless there was some code relying on a plain wrong behavior.

**Comment by elpaso** [comment]
sorry, I meant backport to 3.10 only.

**Comment by elpaso** [comment]
@uclaros thank you!

**Comment by uclaros** [comment]
@elpaso You're welcome :)

**Pull request merged**